pipeline/transformer/transformer_calm/src/main/scala/uk/ac/wellcome/platform/transformer/calm/TransformerWorker.scala

```diff
@@ -37,60 +43,64 @@ trait TransformerWorker[In, SenderDest] {
-    source: Source[StreamMessage, NotUsed]): Source[Unit, NotUsed] =
-    source
-      .via(divertEither(decodeMessage, to = errorSink))
-      .via(divertEither(work, to = errorSink))
-      .via(divertEither(done, to = errorSink))
+  def process(message: NotificationMessage): Result[Unit] = {
+    for {
+      key <- decodeKey(message)
+      recordAndKey <- getRecord(key)
+      work <- work(recordAndKey._1, key)
+      done <- done(work, key)
+    } yield done
+  }
 
-  private def divertEither[I, O](
-    flow: Flow[I, Result[O], NotUsed],
-    to: Sink[Result[_], Future[Done]]): Flow[I, O, NotUsed] =
-    flow.divertTo(errorSink, _.isLeft).collect {
-      case Right(out) => out
+  private def work(sourceData: In, key: StoreKey): Result[TransformedBaseWork] =
+    transformer(sourceData, key.version) match {
+      case Left(err)     => Left(TransformerError(err.toString, sourceData, key))
+      case Right(result) => Right(result)
     }
 
-  val decodeMessage: Flow[StreamMessage, Result[(In, StoreKey)], NotUsed] =
-    Flow.fromFunction(message => decodeKey(message._2) flatMap getRecord)
-
-  val work: Flow[(In, StoreKey), Result[TransformedBaseWork], NotUsed] =
-    Flow.fromFunction {
-      case (sourceData, key) =>
-        transformer(sourceData, key.version) match {
-          case Left(err)     => Left(TransformerError(err.toString))
-          case Right(result) => Right(result)
-        }
+  private def done(work: TransformedBaseWork, key: StoreKey): Result[Unit] =
+    sender.sendT(work) toEither match {
+      case Left(err) => Left(MessageSendError(err.toString, work, key))
+      case Right(_)  => Right((): Unit)
     }
 
-  val done: Flow[TransformedBaseWork, Result[Unit], NotUsed] =
-    Flow.fromFunction(work =>
-      sender.sendT(work) toEither match {
-        case Left(err) => Left(MessageSendError(err.toString))
-        case Right(_)  => Right((): Unit)
-    })
-
-  private def decodeKey(message: NotificationMessage) =
+  private def decodeKey(message: NotificationMessage): Result[StoreKey] =
     fromJson[StoreKey](message.body).toEither match {
-      case Left(err)     => Left(DecodeKeyError(err.toString))
+      case Left(err)     => Left(DecodeKeyError(err.toString, message))
       case Right(result) => Right(result)
     }
 
-  private def getRecord(key: StoreKey) = store.get(key) match {
-    case Left(err)                     => Left(StoreReadError(err.toString))
-    case Right(Identified(key, entry)) => Right((entry, key))
-  }
+  private def getRecord(key: StoreKey): Result[(In, StoreKey)] =
+    store.get(key) match {
+      case Left(err)                     => Left(StoreReadError(err.toString, key))
+      case Right(Identified(key, entry)) => Right((entry, key))
+    }
 
   def run(): Future[Done] =
     stream.runStream(
       name,
       source => {
-        val end = source.via(Flow.fromFunction(message => message._1))
-        val processed = withSource(source)
+        source.mapAsync(2) {
```

Comment:
Probably want this as a parameter on the trait?

Reply:
Why? We do that elsewhere but never use it?

Reply:
It feels presumptuous to assume 2 concurrent transformations will always be the ideal value for any transformer. And this isn't always the case anyway (at least in the METS transformer this isn't true).

How about we set this as to 2 by default and then can be overriden if required?

```scala
val concurrentTransformations: Int = 2
```

Reply:
Was playing it way to much in the non-abstract thinking way forgetting this will be used elsewhere.

Reply:
After having done that though, I wonder if it's a parameter sent to the `run` method?

Reply:
I would probably keep it at trait level, as the caller shouldn't be so concerned about it (seems to be an implementation detail the worker would know more about than whoever is calling the worker). But up to you, would work as a default arg too.